# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [53]:
# Libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [7]:
# your code here
diabetes= datasets.load_diabetes()



Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [6]:
# your code here
diabetes

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990842, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06832974, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286377, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04687948,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452837, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00421986,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [8]:
# your code here
print(f'{diabetes.DESCR}')

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

1. There are 11 attributes in the data. First 10 columns are numeric predictive values and Column 11 is a quantitative measure of disease progression one year after baseline.

2. The relation is the 

3. total of 442

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [10]:
# your code here
diabetes.data.shape

(442, 10)

In [11]:
diabetes.target.shape

(442,)

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:


$y = \beta_0 + \beta_1·X_1 + \beta_2·X_2 + ... + \beta_n·X_n + \epsilon$


where:

- $X_1-X_n$: data 
- $\beta_0$: intercept 
- $\beta_1-\beta_n$: coefficients 
- $\epsilon$: error (cannot explained by model)
- $y$: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [13]:
# your code here
from sklearn.linear_model import LinearRegression


#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [15]:
# your code here
X=diabetes.data

y=diabetes.target

diabetes_model = LinearRegression()


#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [21]:
# your code here
diabetes_data_train = X[:-20]
diabetes_data_test = X[-20:]

diabetes_target_train = y[:-20]
diabetes_target_test = y[-20:]



Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [22]:
# your code here

reg = diabetes_model.fit(diabetes_data_train, diabetes_target_train)

print(reg.intercept_, reg.coef_)


152.76430691633442 [ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
 -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
  7.43519617e+02  7.60951722e+01]


#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [23]:
# your code here
prediction = reg.predict(diabetes_data_test)
prediction

array([197.61846908, 155.43979328, 172.88665147, 111.53537279,
       164.80054784, 131.06954875, 259.12237761, 100.47935157,
       117.0601052 , 124.30503555, 218.36632793,  61.19831284,
       132.25046751, 120.3332925 ,  52.54458691, 194.03798088,
       102.57139702, 123.56604987, 211.0346317 ,  52.60335674])

#### Print your `diabetes_target_test` and compare with the prediction. 

In [24]:
# your code here
print(diabetes_target_test)

[233.  91. 111. 152. 120.  67. 310.  94. 183.  66. 173.  72.  49.  64.
  48. 178. 104. 132. 220.  57.]


#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [25]:
# your answer here
r2_score(diabetes_target_test,prediction)

#is not the perfect result, is quite bad.

0.5850753022690575

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    $H_0 : \beta_1 = \beta_2 = ... = \beta_{10} = 0$

    $H_a$ : At least one $\beta_j \ne 0$ (for j = 1, 2, ..., 10)
    
    
* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

<div style="text-align:center"><img src="../images/statsmodels.png" width=600/></div>

In [ ]:
# your code here

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

In [ ]:
# your answer here

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [26]:
# your code here

auto = pd.read_csv('../data/auto-mpg.csv')
auto


mpg  cylinders  displacement  horse_power  weight  acceleration  \
0    18.0          8         307.0        130.0    3504          12.0   
1    15.0          8         350.0        165.0    3693          11.5   
2    18.0          8         318.0        150.0    3436          11.0   
3    16.0          8         304.0        150.0    3433          12.0   
4    17.0          8         302.0        140.0    3449          10.5   
..    ...        ...           ...          ...     ...           ...   
393  27.0          4         140.0         86.0    2790          15.6   
394  44.0          4          97.0         52.0    2130          24.6   
395  32.0          4         135.0         84.0    2295          11.6   
396  28.0          4         120.0         79.0    2625          18.6   
397  31.0          4         119.0         82.0    2720          19.4   

     model_year                       car_name  
0            70  \t"chevrolet chevelle malibu"  
1            70          \t"buick skylark 320"  
2            70         \t"plymouth satellite"  
3            70              \t"amc rebel sst"  
4            70                \t"ford torino"  
..          ...                            ...  
393          82            \t"ford mustang gl"  
394          82                  \t"vw pickup"  
395          82              \t"dodge rampage"  
396          82                \t"ford ranger"  
397          82                 \t"chevy s-10"  

[398 rows x 8 columns]

Look at the first 5 rows using the `head()` function:

In [27]:
# your code here
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [28]:
# your code here
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 8 columns):
mpg             398 non-null float64
cylinders       398 non-null int64
displacement    398 non-null float64
horse_power     392 non-null float64
weight          398 non-null int64
acceleration    398 non-null float64
model_year      398 non-null int64
car_name        398 non-null object
dtypes: float64(4), int64(3), object(1)
memory usage: 25.0+ KB


What is the newest model year and the oldest model year?

In [45]:
auto.loc[[auto.model_year.idxmin()]]


mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"

In [44]:
auto.loc[[auto.model_year.idxmax()]]


mpg  cylinders  displacement  horse_power  weight  acceleration  \
367  28.0          4         112.0         88.0    2605          19.6   

     model_year                car_name  
367          82  \t"chevrolet cavalier"

Check the dataset for missing values and remove all rows containing at least one missing value.

In [47]:
# check the nan values:
auto.isna().sum()
# horse power has 6, so drop them:
auto.dropna(inplace=True)


Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [48]:
# your code here
auto.cylinders.value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [49]:
# your code here
auto.drop(['car_name'],axis=1,inplace=True)


In [50]:
auto_x = auto.drop('mpg',axis=1)

In [51]:
auto_y=auto.mpg

In [54]:
X_train, X_test, y_train, y_test = train_test_split(auto_x, auto_y, test_size=0.20, random_state=42)



Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [56]:
# your code here

auto_model = LinearRegression().fit(X_train,y_train)

auto_model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)


# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](https://en.wikipedia.org/wiki/Coefficient_of_determination). Its formula is:

<div style="text-align:center"><img src="../images/r-squared.png" width=300/></div>

where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [61]:
# your code here

y_pred = auto_model.predict(X_train)

r2_score(y_train,y_pred)



0.8107227953093896

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [65]:
# your code here

auto_model_test = LinearRegression().fit(X_test,y_test)

y_test_pred = auto_model_test.predict(X_test)

r2_score(y_test_pred,y_test)


0.7839470074716337

#### Explaining the results

Please, compare the results for the training and test sets and comment. What can you say?

We can observe that both results are quite accurate, with a diff of 0.02. both of them with a high r2 score result, which we can say the accuracy is quite low.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [66]:
# your code here

X_train09, X_test09, y_train09, y_test09 = train_test_split(auto_x, auto_y, test_size=0.10, random_state=42)


Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [69]:
# your code here
auto_model09 = LinearRegression().fit(X_train09, y_train09)

Compute the predicted values and r squared score for our new model and new sample data.

In [77]:
# your code here
predict_09 = auto_model09.predict(X_train09)

r2_09 = r2_score( predict_09, y_train09)

print(predict_09, r2_09)

[12.72074548 27.64721889 34.52899085 26.33114709 11.41315151  9.6632474
 17.67126194 31.18298969 25.395304   30.17477011 11.75964595 25.14224248
 24.55764053 33.35989966 28.84013339 17.86976362 20.55806253 23.12836701
 23.11520286 24.61450015  8.3035313  22.83598613 27.17440751 23.00449792
 28.08831113 28.60422025 26.77402648 29.40991667 21.99315699  9.87473174
 23.81191832 13.63402281 24.89688973 26.85637966 23.96061671 25.77677131
 13.67817775 15.52490993 24.48213532 20.89707204 26.22843333 22.35473963
 12.5240259  34.4696109  15.28109843 33.8427865  28.06785634 23.92574835
 16.88868509 16.75876855 26.53513323 25.92522548 30.78124675 30.698788
 21.50345417 17.76971879 20.79439067  9.97969458 27.74634606 20.15679188
  9.53690756 16.92911399 10.03548608 26.76874204 30.65668591 33.36833066
 35.18451003 23.76100523 22.93070795 19.8845697  34.59566455 32.49013354
 25.56559314 24.4086788  32.76696341 29.64128885 19.63142633 21.75145507
 25.27717238 21.40614425 30.08560224 20.93840207 19.70

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [76]:
# your code here

predict_y_test09 = auto_model09.predict(X_test09)

r2_score(predict_y_test09, y_test09)


0.8363926533372654

We can observe a small increace in the r2 score, however, it style have the same % or nearly the same for the train and test model with 90% and 10% of the results.

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE.

In [ ]:
# Libraries
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`.

In [ ]:
# your code here

Fit the model and print the ranking.

In [ ]:
# your code here

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [ ]:
# your code here

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [ ]:
# your code here

#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.